rfc/rfc-0000-common-blueprint.md

```diff
@@ -0,0 +1,220 @@
+As part of this, the RFC proposes that outputs be dynamically typed via a new CR `ClusterBlueprintType` and remove the statically typed outputs that exist today (url, revision, image, config). This allows us to maintain the contract of swappable templates, without restricting outputs to a predefined set. 
+Additionally, the selectors will be removed from blueprints and instead be specified in a new CR `ClusterBlueprintSelector`.
+
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+
+  - Today we have seven CRDs that are needed to create a supply chain and delivery. To simplify the data model for users, we can
+collapse all seven CRDs into a few common CRDs. Additionally, whereas today the overhead to start reconciling templates would be adding five identical
+reconcilers, this would enable us to have one reconciler for all blueprints and more easily add meaningful data to their statuses.
+
+- What use cases does it support?
+
+  - This would allow for snippets and users to specify any output they want, while maintaining strong typing.
+  - The `ClusterBlueprintSelector` will enable users to separate RBAC between template authors and operators who want to restrict what blueprints can be selected.
+
+- What is the expected outcome?
+
+  - Simplified on-boarding and user experience.
+
+# What it is
+[what-it-is]: #what-it-is
+* `ClusterBlueprints` will encapsulate everything the current templates allow, while adding `spec.resources` to enable defining multiple resources (i.e. snippets).
+* A `ClusterBlueprint` can take one of `spec.resources`, `spec.template`, and `spec.ytt`. It will allow inlining of templates in `spec.resources`.
+* A `ClusterBlueprint` must specify a `blueprintTypeRef` which defines the outputs required from the blueprint. The `ClusterBlueprint`'s `spec.outputs` must match those outputs.
+* A `ClusterBlueprintType` defines only `spec.outputs`, which is an array of output names. By leaving the `name` key in the array, we leave open the possibility to later add `type` to each output, though this RFC currently proposes only supporting string outputs.
+* A `ClusterBlueprintSelector` defines a `spec.blueprintRef`, `spec.selector` (per the v1alpha2 syntax), and `spec.ownerType` to define the type of owner to match against. This type takes an `apiVersion` and `kind` to leave open the possibility to later support other owner types, though this RFC proposes only supporting the existing workload and deliverable types.
+
+Example with resources, inlining, two outputs, and workload selector:
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterBlueprint
+metadata:
+  name: my-blueprint
+spec:
+  blueprintTypeRef:
+    name: my-baz-fizz-type
+  
+  # mutually exclusive - resources, template, ytt 
+  resources:
+    - name: first-resource
+      inputs:
+        - resource: zero-resource
+          output: zoo
```

Comment:
what is this field used for? e.g., when templating something with `ytt`, what's the influence of this? 

e.g., previously when consuming some sources (e.g., `sources: [{name: git-repository, resource: source-provider}]`, we'd be able to do in ytt: `#@ data.values.source.url` or `#@ data.values.sources['git-repository'].url` - how does this new format compare?)

Reply:
So before we introduced `ClusterBlueprintTypes`, an input had three fields, `resource` - name of the resource to get the output, `output` - which output to use from the resource, `name` - a name to  refer to the resource in this resource (swapability). Now, we should be able to get rid out of output and should be the same as before.